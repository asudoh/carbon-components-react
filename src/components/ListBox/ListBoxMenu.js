import React from 'react';
import PropTypes from 'prop-types';
import ListBoxMenuItem from './ListBoxMenuItem';
import childrenOfType from '../../prop-types/childrenOfType';

/**
 * `ListBoxMenu` is a simple container node that isolates the `list-box__menu`
 * class into a single component. It is also being used to validate given
 * `children` components.
 */
const ListBoxMenu = ({ children, prefix, ...rest }) => {
  return (
    <div className={`${prefix}--list-box__menu`} {...rest}>
      {children}
    </div>
  );
};

ListBoxMenu.propTypes = {
  /**
   * Provide the contents of your ListBoxMenu
   */
  children: childrenOfType(ListBoxMenuItem),

  /**
   * The selector prefix
   */
  prefix: PropTypes.string,
};

ListBoxMenu.defaultProps = {
  prefix: 'bx',
};

export default ListBoxMenu;
