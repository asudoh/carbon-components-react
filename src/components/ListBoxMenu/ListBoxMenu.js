import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Icon from '../Icon';

const ListBoxMenu = ({
  items,
  selectedItem,
  highlightedIndex,
  itemToString,
  getItemProps,
}) => (
  <div className="bx--list-box__menu">
    {items.map((item, i) => (
      <div
        key={item.id}
        className={cx({
          'bx--list-box__menu-item': true,
          'bx--list-box__menu-item--highlighted': highlightedIndex === i,
          'bx--list-box__menu-item--active': selectedItem.indexOf(item) !== -1,
        })}
        {...getItemProps({
          item,
          index: i,
        })}>
        <input
          id={item.id}
          name={itemToString(item)}
          className="bx--checkbox"
          type="checkbox"
          checked={selectedItem.indexOf(item) !== -1}
          readOnly={true}
        />
        <label htmlFor={item.id} className="bx--checkbox-label">
          <span className="bx--checkbox-appearance">
            <Icon className="bx--checkbox-checkmark" name="checkmark" />
          </span>
          {itemToString(item)}
        </label>
      </div>
    ))}
  </div>
);

ListBoxMenu.propTypes = {
  items: PropTypes.array.isRequired,
  selectedItem: PropTypes.array.isRequired,
  highlightedIndex: PropTypes.number,
  itemToString: PropTypes.func.isRequired,
  getItemProps: PropTypes.func.isRequired,
};

export default ListBoxMenu;
