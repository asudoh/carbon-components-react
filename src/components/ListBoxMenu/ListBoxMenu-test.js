import React from 'react';
import { shallow } from 'enzyme';
import ListBoxMenu from '../ListBoxMenu';
import {
  generateItems,
  generateGenericItem,
} from '../../tools/testing/items';

describe('ListBoxMenu', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      itemToString: jest.fn(({ label }) => label),
      getItemProps: jest.fn(() => ({})),
    };
  });

  it('should render', () => {
    const wrapper = shallow(
      <ListBoxMenu
        items={generateItems(5, generateGenericItem)}
        selectedItem={[]}
        highlightedIndex={null}
        {...mockProps}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call `getItemProps` for each item that is generated', () => {
    const itemsCount = 5;
    shallow(
      <ListBoxMenu
        items={generateItems(itemsCount, generateGenericItem)}
        selectedItem={[]}
        highlightedIndex={null}
        {...mockProps}
      />
    );
    expect(mockProps.getItemProps).toHaveBeenCalledTimes(itemsCount);
  });

  describe('with selected items', () => {
    it('should add an active class to selected items', () => {
      const items = generateItems(5, generateGenericItem);
      const wrapper = shallow(
        <ListBoxMenu
          items={items}
          selectedItem={[items[0], items[1]]}
          highlightedIndex={null}
          {...mockProps}
        />
      );
      expect(wrapper.find('.bx--list-box__menu-item--active').length).toBe(2);
    });

    it('should set the checkbox value to checked for a selected item', () => {
      const items = generateItems(5, generateGenericItem);
      const wrapper = shallow(
        <ListBoxMenu
          items={items}
          selectedItem={[items[0], items[1]]}
          highlightedIndex={null}
          {...mockProps}
        />
      );
      wrapper.find('.bx--list-box__menu-item--active input').forEach(node => {
        expect(node.prop('checked')).toBe(true);
      });
    });
  });

  describe('with highlighted index', () => {
    it('should add a highlighted class for the specific item at the index', () => {
      const wrapper = shallow(
        <ListBoxMenu
          items={generateItems(5, generateGenericItem)}
          selectedItem={[]}
          highlightedIndex={0}
          {...mockProps}
        />
      );
      expect(wrapper.find('.bx--list-box__menu-item--highlighted').length).toBe(
        1
      );
    });
  });
});
