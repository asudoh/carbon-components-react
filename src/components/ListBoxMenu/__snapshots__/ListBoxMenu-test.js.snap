// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxMenu should render 1`] = `
<div
  className="bx--list-box__menu"
>
  <div
    className="bx--list-box__menu-item"
    key="id-0"
  >
    <input
      checked={false}
      className="bx--checkbox"
      id="id-0"
      name="Item 0"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="bx--checkbox-label"
      htmlFor="id-0"
    >
      <span
        className="bx--checkbox-appearance"
      >
        <Icon
          className="bx--checkbox-checkmark"
          description="Provide a description that will be used as the title"
          fillRule="evenodd"
          name="checkmark"
          role="img"
        />
      </span>
      Item 0
    </label>
  </div>
  <div
    className="bx--list-box__menu-item"
    key="id-1"
  >
    <input
      checked={false}
      className="bx--checkbox"
      id="id-1"
      name="Item 1"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="bx--checkbox-label"
      htmlFor="id-1"
    >
      <span
        className="bx--checkbox-appearance"
      >
        <Icon
          className="bx--checkbox-checkmark"
          description="Provide a description that will be used as the title"
          fillRule="evenodd"
          name="checkmark"
          role="img"
        />
      </span>
      Item 1
    </label>
  </div>
  <div
    className="bx--list-box__menu-item"
    key="id-2"
  >
    <input
      checked={false}
      className="bx--checkbox"
      id="id-2"
      name="Item 2"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="bx--checkbox-label"
      htmlFor="id-2"
    >
      <span
        className="bx--checkbox-appearance"
      >
        <Icon
          className="bx--checkbox-checkmark"
          description="Provide a description that will be used as the title"
          fillRule="evenodd"
          name="checkmark"
          role="img"
        />
      </span>
      Item 2
    </label>
  </div>
  <div
    className="bx--list-box__menu-item"
    key="id-3"
  >
    <input
      checked={false}
      className="bx--checkbox"
      id="id-3"
      name="Item 3"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="bx--checkbox-label"
      htmlFor="id-3"
    >
      <span
        className="bx--checkbox-appearance"
      >
        <Icon
          className="bx--checkbox-checkmark"
          description="Provide a description that will be used as the title"
          fillRule="evenodd"
          name="checkmark"
          role="img"
        />
      </span>
      Item 3
    </label>
  </div>
  <div
    className="bx--list-box__menu-item"
    key="id-4"
  >
    <input
      checked={false}
      className="bx--checkbox"
      id="id-4"
      name="Item 4"
      readOnly={true}
      type="checkbox"
    />
    <label
      className="bx--checkbox-label"
      htmlFor="id-4"
    >
      <span
        className="bx--checkbox-appearance"
      >
        <Icon
          className="bx--checkbox-checkmark"
          description="Provide a description that will be used as the title"
          fillRule="evenodd"
          name="checkmark"
          role="img"
        />
      </span>
      Item 4
    </label>
  </div>
</div>
`;
